### ISSUE TITLE
NFS Server resources disappear following ETCD restart.

### ISSUE THREAD

**Issue by lcooper40** [rook/rook#4142]
**Is this a bug report or feature request?**
* Bug Report

**Deviation from expected behavior:**
Restarting ETCD in the k8s cluster appears to cause resources to disappear and nfs pod terminates without coming back.
Everything that has been developed deploys and works. I am able to write to the NFS share thats created and everything looks good up until the point ETCD is restarted.
Upon ETCD restarting the statefulset, configmap and controllerrevision resources disappear and the NFS pod goes into terminating. What is left behind is the nfsserver resource and the PV's/PVC's and storage class.

**Expected behavior:**
To be able to restart the k8s components (etcd) and the NFS pod recreate itself successfully.

**How to reproduce it (minimal and precise):**
Create a nfs server which is backed by local storage and restart ETCD in the cluster.

The mapping in my env is:
PV <-- PVC <-- NFS <-- NFS Storage Class <-- NFS PVC

**File(s) to submit**:
PV.yaml
apiVersion: v1
kind: PersistentVolume
metadata:
  name: logstash-pv
  namespace: rook-nfs
  labels:
    type: local
spec:
  storageClassName: local-storage
  capacity:
    storage: 10Gi
  accessModes:
    - ReadWriteOnce
  hostPath:
    path: "/mnt/disks/nfs/logstash"
  claimRef:
    name: logstash-pvc

PVC.yaml
apiVersion: v1
kind: PersistentVolumeClaim
metadata:
  name: logstash-pvc
  namespace: rook-nfs
spec:
  storageClassName: local-storage
  accessModes:
    - ReadWriteOnce
  resources:
    requests:

NFS.yaml
apiVersion: nfs.rook.io/v1alpha1
kind: NFSServer
metadata:
  name: logstash-nfs
  namespace: rook-nfs
spec:
  replicas: 1
  exports:
  - name: logstash
    server:
      accessMode: ReadWrite
      squash: "none"
    persistentVolumeClaim:
      claimName: logstash-pvc

NFS_SC.yaml
apiVersion: storage.k8s.io/v1
kind: StorageClass
metadata:
  name: logstash-nfs-share
  labels:
    app: logstash-nfs
    name: logstash-nfs-share
parameters:
  exportName: logstash
  nfsServerName: logstash-nfs
  nfsServerNamespace: rook-nfs
provisioner: rook.io/nfs-provisioner
reclaimPolicy: Delete
volumeBindingMode: Immediate

NFS_PVC.yaml
apiVersion: v1
kind: PersistentVolumeClaim
metadata:
  name: logstash-share-pvc
  namespace: rook-nfs
spec:
  storageClassName: "logstash-nfs-share"
  accessModes:
    - ReadWriteMany
  resources:
    requests:
      storage: 10Gi

The logs from the pod show nothing in relation to the error. The below are the events from describing the pod just before it is permanently terminated.

Events:
  Type     Reason          Age              From                                    Message
  ----     ------          ----             ----                                    -------
  Normal   Scheduled       19m              default-scheduler                       Successfully assigned rook-nfs/logstash-nfs-0 to ip-*.ec2.internal
  Normal   Pulled          19m              kubelet, ip-*.ec2.internal  Container image "rook/nfs:master" already present on machine
  Normal   Created         19m              kubelet, ip-*.ec2.internal  Created container
  Normal   Started         19m              kubelet, ip-*.ec2.internal  Started container
  Normal   SandboxChanged  33s              kubelet, ip-*.ec2.internal  Pod sandbox changed, it will be killed and re-created.
  Normal   Pulled          32s              kubelet, ip-*.ec2.internal  Container image "rook/nfs:master" already present on machine
  Normal   Created         32s              kubelet, ip-*.ec2.internal  Created container
  Normal   Started         32s              kubelet, ip-*.ec2.internal  Started container
  Normal   Killing         5s               kubelet, ip-*.ec2.internal  Killing container with id docker://logstash-nfs:Need to kill Pod
  Warning  FailedMount     1s (x3 over 4s)  kubelet, ip-*.ec2.internal  MountVolume.SetUp failed for volume "logstash-nfs" : configmaps "logstash-nfs" not found

**Environment**:
* OS (e.g. from /etc/os-release): Centos7
* Kernel (e.g. `uname -a`):  3.10.0-957.el7.x86_64
* Cloud provider or hardware configuration: AWS
* Rook version (use `rook version` inside of a Rook Pod): v1.1.0-beta.0.299.g1081c58
* Storage backend version (e.g. for ceph do `ceph -v`): local-storage
* Kubernetes version (use `kubectl version`): 1.13.8
* Kubernetes cluster type (e.g. Tectonic, GKE, OpenShift): Kubespray
* Storage backend status (e.g. for Ceph use `ceph health` in the [Rook Ceph toolbox](https://rook.io/docs/rook/master/ceph-toolbox.html)): local-storage

**Comment by lcooper40** [comment]
Hi, 

Has anyone had a chance to pick this up and replicate the same issue ?

Thanks

**Comment by lcooper40** [comment]
@travisn 

I have used the latest image (v1.1.0-beta.0.326.g0178724) and I still see the same issue. On restart of ETCD the resources are removed.
Provided below are the logs for the operator showing it bringing the resources up no problem untill the restart of etcd and then it gets into trouble trying to start a new stateful set.

```
2019-10-24 07:56:00.597214 I | rookcmd: starting Rook v1.1.0-beta.0.326.g0178724 with arguments '/usr/local/bin/rook nfs operator'
2019-10-24 07:56:00.597333 I | rookcmd: flag values: --add_dir_header=false, --alsologtostderr=false, --help=false, --log-flush-frequency=5s, --log-level=INFO, --log_backtrace_at=:0, --log_dir=, --log_file=, --log_file_max_size=1800, --logtostderr=true, --operator-image=, --service-account=, --skip_headers=false, --skip_log_headers=false, --stderrthreshold=2, --v=0, --vmodule=
2019-10-24 07:56:00.597341 I | nfscmd: starting NFS operator
2019-10-24 07:56:00.618935 I | nfs-operator: start watching nfs server resources in namespace
2019-10-24 07:56:00.618974 I | nfs-operator: Started watch for NFS Servers
2019-10-24 07:56:09.249822 I | nfs-operator: new NFS server logstash-nfs added to namespace rook-nfs
2019-10-24 07:56:09.249841 I | nfs-operator: validating nfs server spec in namespace rook-nfs
2019-10-24 07:56:09.249848 I | nfs-operator: creating nfs server service in namespace rook-nfs
2019-10-24 07:56:09.266363 I | nfs-operator: nfs service logstash-nfs started in namespace rook-nfs
2019-10-24 07:56:09.266389 I | nfs-operator: creating nfs server configuration in namespace rook-nfs
2019-10-24 07:56:09.277610 I | nfs-operator: creating nfs server stateful set in namespace rook-nfs
2019-10-24 07:56:09.303515 I | nfs-operator: stateful set logstash-nfs created in namespace rook-nfs
2019-10-24 07:56:44.505958 I | nfs-operator: cluster logstash-nfs deleted from namespace rook-nfs

**Restarted ETCD at this point**
2019-10-24 07:57:07.971087 I | nfs-operator: new NFS server logstash-nfs added to namespace rook-nfs
2019-10-24 07:57:07.971116 I | nfs-operator: validating nfs server spec in namespace rook-nfs
2019-10-24 07:57:07.971125 I | nfs-operator: creating nfs server service in namespace rook-nfs
2019-10-24 07:57:08.000071 I | nfs-operator: nfs service logstash-nfs started in namespace rook-nfs
2019-10-24 07:57:08.000088 I | nfs-operator: creating nfs server configuration in namespace rook-nfs
2019-10-24 07:57:08.006361 I | nfs-operator: creating nfs server stateful set in namespace rook-nfs
2019-10-24 07:57:08.012333 I | nfs-operator: stateful set logstash-nfs created in namespace rook-nfs
```

**Comment by travisn** [comment]
@lcooper40 Found the issue, #4179 will fix it...

**Comment by lcooper40** [comment]
This hasn't fixed my issue. The pod still terminates following a restart. Operator log below just to show I'm using the newest version which I assume has the updated code.

2019-10-25 10:41:20.932783 I | rookcmd: starting Rook v1.1.0-beta.0.336.g1f78304 with arguments '/usr/local/bin/rook nfs operator'
2019-10-25 10:41:20.932929 I | rookcmd: flag values: --add_dir_header=false, --alsologtostderr=false, --help=false, --log-flush-frequency=5s, --log-level=INFO, --log_backtrace_at=:0, --log_dir=, --log_file=, --log_file_max_size=1800, --logtostderr=true, --operator-image=, --service-account=, --skip_headers=false, --skip_log_headers=false, --stderrthreshold=2, --v=0, --vmodule=
2019-10-25 10:41:20.932943 I | nfscmd: starting NFS operator
2019-10-25 10:41:20.971808 I | nfs-operator: start watching nfs server resources in namespace
2019-10-25 10:41:20.971858 I | nfs-operator: Started watch for NFS Servers
2019-10-25 10:41:21.004796 I | nfs-operator: new NFS server logstash-nfs added to namespace rook-nfs
2019-10-25 10:41:21.004829 I | nfs-operator: validating nfs server spec in namespace rook-nfs
2019-10-25 10:41:21.004837 I | nfs-operator: creating nfs server service in namespace rook-nfs
2019-10-25 10:41:21.048863 I | nfs-operator: nfs service logstash-nfs already exists in namespace rook-nfs
2019-10-25 10:41:21.048895 I | nfs-operator: creating nfs server configuration in namespace rook-nfs
2019-10-25 10:41:21.056404 E | nfs-operator: Unable to create NFS ConfigMap configmaps "logstash-nfs" already exists
2019-10-25 10:41:21.056431 I | nfs-operator: creating nfs server stateful set in namespace rook-nfs
2019-10-25 10:41:21.081864 I | nfs-operator: stateful set logstash-nfs already exists in namespace rook-nfs

**Comment by lcooper40** [comment]
@leseb  @travisn 

We have since tried this version 
v1.1.0.344.g30d019a

This was successful and the resources recreated successfully after restart k8s. Will this be built into a full release version ?

**Comment by travisn** [comment]
@lcooper40 Great to hear it's working for you. Yes, this will be included in v1.1.5 when it is released. It could be a week or two depending on urgency of the release.
